src/request_executor.cc

```diff
@@ -56,12 +56,110 @@ void
+                sname, dims_vector, datatype, memory_type, memory_type_id,
+                const_cast<void*>(base), byte_size,
+                nullptr /* DLManagedTensor */);
+
+            // Load the data so that it is deallocated automatically.
+            std::unique_ptr<PbMemory> pb_memory(
+                reinterpret_cast<PbMemory*>(userp));
+            pb_tensor->SetMemory(std::move(pb_memory));
+            output_tensors.push_back(pb_tensor);
+          } else {
+            output_tensors.push_back(std::make_shared<PbTensor>(
+                sname, dims_vector, datatype, memory_type, memory_type_id,
+                const_cast<void*>(base), byte_size,
+                nullptr /* DLManagedTensor */));
+          }
+        } else {
+          output_tensors.push_back(std::make_shared<PbTensor>(
+              sname, dims_vector, datatype, memory_type, memory_type_id,
+              const_cast<void*>(base), byte_size,
+              nullptr /* DLManagedTensor */));
+        }
+      }
+      std::shared_ptr<PbError> pb_error;
+      infer_response =
+          std::make_unique<InferResponse>(output_tensors, pb_error);
+    }
+    catch (const PythonBackendException& pb_exception) {
+      if (response != nullptr) {
+        LOG_IF_ERROR(
+            TRITONSERVER_InferenceResponseDelete(response),
+            "Failed to delete inference response.");
+
+        response = nullptr;
+      }
+      std::shared_ptr<PbError> pb_error =
+          std::make_shared<PbError>(pb_exception.what());
+      infer_response = std::make_unique<InferResponse>(
+          std::vector<std::shared_ptr<PbTensor>>{}, pb_error);
+    }
+
+    infer_response->SetCompletedResponse(response);
+
+    if (!(*p)->IsDecoupled()) {
+      infer_response->ResetNextResponseFuture();
+      (*p)->SetValueForPrevPromise(std::move(infer_response));
+      (*p)->ResetPrevPromise();
+    } else {
+      if ((flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) == 0) {
+        // Not the last reponse. Need to store the promise associated with the
+        // next future.
+        auto promise = new std::promise<std::unique_ptr<InferResponse>>();
+        infer_response->SetNextResponseFuture(promise);
+        (*p)->SetValueForPrevPromise(std::move(infer_response));
+        (*p)->SetPrevPromise(&promise);
+      } else {
+        // The last response.
+        infer_response->ResetNextResponseFuture();
+        (*p)->SetValueForPrevPromise(std::move(infer_response));
+        (*p)->ResetPrevPromise();
+      }
+    }
+  } else if (
+      (*p)->IsDecoupled() &&
+      (flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) != 0) {
+    // An empty response may be the last reponse for decoupled models.
+    (*p)->SetValueForPrevPromise(std::unique_ptr<InferResponse>{});
+    (*p)->ResetPrevPromise();
+  } else {
+    (*p)->SetValueForPrevPromise(std::unique_ptr<InferResponse>{});
+    (*p)->ResetPrevPromise();
+    throw PythonBackendException("Unexpected empty response.");
```

Comment:
I don't think you can throw an exception here.

Reply:
Great catch. Updated to make the error being returned as a part of  `InferResponse` object rather than throwing exception here.